keps/sig-network/NNNN-nftables-proxy/README.md

```diff
@@ -0,0 +1,1342 @@
+kernel API is mostly dead.
+
+However, this new mode does not add any new _syntax_, and so it is not
+possible to use any of the new nftables features (like maps) that are
+not present in iptables.
+
+Furthermore, the compatibility constraints imposed by the user-facing
+API of the iptables binaries themselves prevent them from being able
+to take advantage of many of the performance improvements associated
+with nftables.
+
+### The `iptables` mode of kube-proxy has grown crufty
+
+Because `iptables` is the default kube-proxy mode, it is subject to
+strong backward-compatibility constraints which mean that certain
+"features" that are now considered to be bad ideas cannot be removed
+because they might break some existing users. A few examples:
+
+  - It allows NodePort services to be accessed on `localhost`, which
+    requires it to set a sysctl to a value that may introduce security
+    holes on the system. More generally, it defaults to having
+    NodePort services be accessible on _all_ node IPs, when most users
+    would probably prefer them to be more restricted.
+
+  - It implements the `LoadBalancerSourceRanges` feature for traffic
+    addressed directly to LoadBalancer IPs, but not for traffic
+    redirected to a NodePort by an external LoadBalancer.
+
+  - Some new functionality only works correctly if the administrator
+    passes certain command-line options to kube-proxy (eg,
+    `--cluster-cidr`), but we cannot make those options be mandatory,
+    since that would break old clusters that aren't passing them.
+
+A new kube-proxy, which existing users would have to explicitly opt
+into, could revisit these and other decisions.
+
+### We will hopefully be able to trade 2 supported backends for 1
+
+Right now SIG Network is supporting both the `iptables` and `ipvs`
+backends of kube-proxy, and does not feel like it can ditch `ipvs`
+because of performance issues with `iptables`. If we create a new
+backend which is as functional and non-buggy as `iptables` but as
+performant as `ipvs`, then we could (eventually) deprecate both of the
+existing backends and only have one backend to support in the future.
+
+### Writing a new kube-proxy mode may help with our "KPNG" goals
+
+The [KPNG] (Kube-Proxy Next Generation) working group has been working
+on the future of kube-proxy's underlying architecture. They have
+recently proposed a [kube-proxy library KEP]. Creating a new proxy
+mode which will be officially supported, but which does not (yet) have
+the same compatibility and non-bugginess requirements as the
+`iptables` and `ipvs` modes should help with that project, because we
+can target the new backend to the new library without worrying about
+breaking the old backends.
+
+[KPNG]: https://github.com/kubernetes-sigs/kpng
+[kube-proxy library KEP]: https://github.com/kubernetes/enhancements/pull/3649
+
+### Goals
+
+- Design and implement an `nftables` mode for kube-proxy.
+
+    - Drop support for localhost nodeports
```

Comment:
Yeah. So this KEP only covers adding the new mode, not changing the default, but we would have to eventually figure that out, if we are going to do that.

Reply:
I haven't investigated the details of exactly what makes the route_localnet hack necessary and how it could be avoided. My impression though was that we think that localhost nodeports are not commonly used, and were never really _intended_ to be something that worked, so we kind of want to get rid of them?

We absolutely _can_ have them still work if we want to; I was just assuming we wanted to take advantage of the opportunity to kill them.

Reply:
I think it's worth discussing.  Clearly I *did* intend them to work (or I would not have done the regrettable route_localnet thing).  It's necessary because the kernel assumes it can not route local traffic between interfaces, so it just drops it.  What we really wanted (I think) was to recognize it as local (locally originated, not routed) and then allow NAT on it, but it seems to be all or nothing.  Route it arbitrarily or not at all.

The discussion then is about whether my intention was misguided or useful, and whether anyone would ACTUALLY miss having it.  I only know of that one use case (localhost proxy to registry)